### PR TITLE
Add metric on datapath update latency due to FQDN IP updates

### PR DESCRIPTION
**Pull request by rahulkjoshi**
While there is an overall processing metric, it would be good to extract the datapath update time from other sections of the FQDN processing. This makes it easier to diagnose the source of any backups.

It should be noted that dataplaneTime will be capped to `option.Config.FQDNProxyResponseMaxDelay`. After that time has elapsed, the update is cancelled and the DNS packet forwarded to the endpoint.

```release-note
Add metric on datapath update latency due to FQDN IP updates
```

Signed-off-by: Rahul Joshi <rkjoshi@google.com>

### CODE REVIEW

**Review by tommyp1ckles — APPROVED** [review]
This looks good to me, @christarazi @aanm Can one of you take a look as well.

**Review by aanm — APPROVED** [review]

File: daemon/cmd/fqdn.go
Comment (christarazi) [inline daemon/cmd/fqdn.go]: Shouldn't `stat.DataplaneTime.End(true)` be inside `endMetric`?

**Review by christarazi — CHANGES_REQUESTED** [review]
Overall changes LGTM. One question about the implementation.

A couple of notes:
* This metric will be capped by `FQDNProxyResponseMaxDelay`, just FYI. I think it's worth mentioning in the commit msg.
* Speaking of commit msgs, could you add the PR description into the commit msg? This is so that we have the context inside git history when looking at `git log` in the future.

File: daemon/cmd/fqdn.go
Comment (rahulkjoshi) [reply daemon/cmd/fqdn.go]: I decided not to since `endMetric` gets called from atleast one other place (see line 457). I wasn't sure what happens if you call end on a span that was never started.

I'm ok either way.

Comment (rahulkjoshi) [reply daemon/cmd/fqdn.go]: I decided not to since `endMetric` gets called from atleast one other place (see line 457). I wasn't sure what happens if you call end on a span that was never started.

I'm ok either way.

Comment (christarazi) [reply daemon/cmd/fqdn.go]: Looking at the code, it seems to be supported and the metric will just be zero-valued anyway, so let's move it there, so we keep the code in a consistent pattern. Thanks.

Comment (rahulkjoshi) [reply daemon/cmd/fqdn.go]: done.

Comment (christarazi) [reply daemon/cmd/fqdn.go]: I'm not seeing the change actually

Comment (rahulkjoshi) [reply daemon/cmd/fqdn.go]: 🤦🏽 I guess it helps if you actually stage your changes.

Should be better now

**Review by christarazi — APPROVED** [review]

**Review by tommyp1ckles — APPROVED** [review]

**Comment by christarazi** [comment]
Marking ready to merge as we have known CI failures and code approval.

By the way Rahul, `/test` is only allowed for Cilium contributors who already have write access, so it won't work for other contributors. :)

**Comment by joamaki** [comment]
This is erroring in the prometheus metrics check, but cause seems to be an old `tests-smoke.yaml` that isn't filtering out the `go_` metrics. Master's `tests-smoke.yaml` does filter them, so ignoring the failure. There's also a flaky test in `test-runtime`, but it's unrelated. Merging.

**Comment by christarazi** [comment]
Nominating for v1.11 and v1.10 backports ala https://github.com/cilium/cilium/pull/20225#issuecomment-1158076444

File: daemon/cmd/fqdn.go
Comment (joestringer) [inline daemon/cmd/fqdn.go:520]: While reviewing these metrics in a live cluster, I found that this is including a bunch of control plane processing time since it includes the full `UpdateGenerateDNS`, `DNSHistory.Update()` etc. So this includes all of the policy processing time as well. I think this is a bit misleading compared to the name 'dataplaneTime'.

@rahulkjoshi Have you been able to deploy this and observe the metrics / draw conclusions from these?

Comment (rahulkjoshi) [reply daemon/cmd/fqdn.go:520]: I haven't had a chance to deploy this and really play with it yet. 

Do you think a rename might help capture what's happening here? I expect the actual datapath update + updating the DNS caches to be the parts which hit bottlenecks at scale/in certain configurations. I was hoping this might get us a read on that.

Comment (joestringer) [reply daemon/cmd/fqdn.go:520]: Hmm, it's a good question. If we break down the entire lifetime of the DNS request+response stream, we have something like this:

```mermaid
sequenceDiagram
    participant [@client]
    participant DNS Proxy
    participant enforceConcurrencyLimit()
    participant CheckAllowed()
    participant [@upstream dns]
    participant ProxyLogger
    participant NameManager
    participant IdentityAllocator
    participant SelectorCache
    participant PolicyCache
    participant EndpointManager
    participant Endpoint
    participant BPF PolicyMap
    participant BPF IPCache

    [@client]->>+DNS Proxy: Client DNS request
    DNS Proxy->>+enforceConcurrencyLimit(): Check ratelimit
    Note over DNS Proxy,enforceConcurrencyLimit(): semaphoreTime
    enforceConcurrencyLimit()->>+DNS Proxy: Proceed
    Note over DNS Proxy,CheckAllowed(): Start(processingTime)
    DNS Proxy->>+CheckAllowed(): Enforce L7 DNS policy
    Note over CheckAllowed(): policyCheckTime
    CheckAllowed()->>+DNS Proxy: Allow
    Note over DNS Proxy,CheckAllowed(): Stop(processingTime)

    DNS Proxy->>+[@upstream dns]: Upstream DNS request
    Note over DNS Proxy,[@upstream dns]: upstreamTime
    [@upstream dns]->>+DNS Proxy: Upstream DNS response

    Note over DNS Proxy,BPF IPCache: Start(processingTime)
    DNS Proxy->>+ProxyLogger: Generate L7 monitor event
    Note over DNS Proxy,BPF IPCache: Start(dataplaneTime)
    DNS Proxy->>+NameManager: Calculate policy implications
    NameManager->>+IdentityAllocator: Locally allocate identities for the IPs
    NameManager->>+SelectorCache: Associate new identities with the relevant selectors
    SelectorCache->>+PolicyCache: Store the identities in the per-identity policy caches
    NameManager->>+EndpointManager: Notify endpoints to update policy maps
    loop Update all Endpoints
    EndpointManager->>+Endpoint: Trigger Policy Updates ("Regenerate")
    Endpoint->>+PolicyCache: Fetch incremental updates
    Endpoint->>+BPF PolicyMap: Synchronize BPF policy maps
    end
    NameManager->>+DNS Proxy: Return allocated identities, waitgroup
    DNS Proxy->>+DNS Proxy: Wait for policy updates
    DNS Proxy->>+BPF IPCache: Synchronize new identities to IPCache
    Note over BPF IPCache,DNS Proxy: Stop(dataplaneTime)
    Note over BPF IPCache,DNS Proxy: Stop(processingTime)

    DNS Proxy->>+[@client]: ClientDNS response
```

Now, the graph there is not scaled to the time spent in each area, but there's a whole bunch of control plane logic to prepare the dataplane which is all currently covered by `dataplaneTime`. Some of those portions would also be covered by the `cilium_policy_regeneration_time_stats_seconds` metric but there's not a super great structure in place in the code to be able to get better insights into the way that the current "dataplaneTime" is spent. I'd expect most of the time to be spent there, and the current metric only serves the purpose of eliminating the impact of the L7 logging.

Comment (rahulkjoshi) [reply daemon/cmd/fqdn.go:520]: What if we move the start of this block:

- Option 1: to line 551 -- Helps capture just the time spent waiting on datapath. Not totally sure what we miss out on capturing
- Option 2: to line 538 -- I'm not sure what the DNSHistory.Update command does on your diagram, but does that move us closer to just "datapath" stuff?

Comment (joestringer) [reply daemon/cmd/fqdn.go:520]: option (1) seems reasonable, though it may be worth adding another metric/span for the `UpdateGenerateDNS` call, maybe something like `policyCalculationTime`?

I actually didn't include `DNSHistory.Update()` in my diagram above as I anticipated that it is a relatively minor contributor to the overall time. I don't know the exact details but I believe that's just updating a cache in the NameManager for IP<->FQDN which is used for performing GC on those name->IP mappings. I expect that most of the time is spent in `UpdateGenerateDNS()`, though that portion also triggers some async datapath policymap update processes. The latter is why we have the waitgroup at the end to ensure those async operations are completed before proceeding to populate the ipcache (which I think should also still be part of the datapath programming piece).

Comment (rahulkjoshi) [reply daemon/cmd/fqdn.go:520]: sgtm --  I can tackle this in a few weeks when I'm less busy (or anyone else is more than welcome to help out)

As an aside, do we have a metric which is just a total latency as observed by any clients of DNS Proxy? It's roughly equal to processingTime + upstreamTime, but that's not something you can compute after the fact in prometheus.

Comment (joestringer) [reply daemon/cmd/fqdn.go:520]: OK thanks! Glancing at `ServeDNS()`, I don't see a holistic metric like that today. I think it'd be reasonable to add a `totalTime` label to the same metric to also measure that.

**Pull request merged**
